src/merkle/mmr/accumulator.rs

```diff
@@ -58,4 +44,46 @@ impl MmrPeaks {
             .merkle_path
             .verify(opening.relative_pos() as u64, value, root)
     }
+
+    /// Flattens and pads the peaks to make hashing inside of the Miden VM easier.
+    ///
+    /// The procedure will:
+    /// - Flatten the vector of Words into a vector of Felts.
+    /// - Pad the peaks with ZERO to an even number of words, this removes the need to handle RPO padding.
+    /// - Pad the peaks to a minimum length of 16 words, which reduces the constant cost of
+    ///   hashing.
+    fn flatten_and_pad_peaks(&self) -> Vec<Felt> {
+        let num_peaks = self.peaks.len();
+
+        // To achieve the padding rules above we calculate the length of the final vector.
+        let len = if num_peaks < 16 {
+            64
+        } else if num_peaks % 2 == 1 {
+            (num_peaks + 1) * 4
+        } else {
+            num_peaks * 4
+        };
+
+        let mut elements = Vec::with_capacity(len);
+        elements.extend_from_slice(&self.peaks.as_slice().concat());
+        elements.resize(len, ZERO);
+        elements
+    }
+}
+
+/// Converts the MMR peaks into a vector of Felts.
+///
+/// The vector of Felts is arranged as follows:
+/// [num_leaves, PEAK_0, PEAK_1, ..., PEAK_N]
+///
+/// num_leaves is contains the number of leaves in the MMR.
+/// PEAK_0 is the 0'th peak in the MMR.
+/// PEAK_1 is the 1'st peak in the MMR.
+/// PEAK_N is the N'th peak in the MMR.
+impl From<&MmrPeaks> for Vec<Felt> {
+    fn from(mmr: &MmrPeaks) -> Self {
+        let mut elements = mmr.flatten_and_pad_peaks();
+        elements.insert(0, Felt::new(mmr.num_leaves as u64));
+        elements
+    }
```

Comment:
I personally prefer to always use named functions instead of `From` generic implementations when there is some embedded logic.

In this case, we pad/truncate the elements count to a certain length, and then we suffix it with `num_leaves`.

Also, it's always best to leave to the user the choice to when alloc. If we opt to collect data, manipulate it, and return a new set, we make the job of the compiler to auto vectorize very hard!

Having that said, I'd change a bit the `flatten_and_pad_peaks` to something like

```rust
fn flatten_and_pad_peaks(&self) -> impl Iterator<Item = &Felt> {
    use crate::WORD_SIZE;
    use core::iter;

    let num_peaks = self.peaks.len();

    // To achieve the padding rules above we calculate the length of the final vector.
    let len = if num_peaks < 16 {
        64
    } else if num_peaks % 2 == 1 {
        (num_peaks + 1) * WORD_SIZE
    } else {
        num_peaks * WORD_SIZE
    };

    let pad = len.saturating_sub(num_peaks * WORD_SIZE);
    self.peaks
        .iter()
        .flatten()
        .take(len)
        .chain(iter::repeat(&ZERO).take(pad))
}
```

I would also be more verbose on the documentation of the picked len values so all the variants there are explicitly described.

And, instead of using a generic `From`, I would have something explicit and documented

```rust
fn to_elements(&self) -> impl Iterator<Item = Felt> + '_ {
    use core::iter;
    iter::once(Felt::new(self.num_leaves as u64)).chain(self.flatten_and_pad_peaks().copied())
}
```

As alternative to that, we could have an element that holds the `Felt` of `self.num_leaves` so we don't need to copy the elements

```rust
fn to_elements(&self) -> impl Iterator<Item = &Felt> {
    use core::iter;
    iter::once(self.num_leaves_element).chain(self.flatten_and_pad_peaks().copied())
}
```

With that, if the user wants a `Vec`, he can just

```rust
fn foo(peaks: &MmrPeaks) -> Vec<Felt> {
    peaks.to_elements().collect()
}
```

Reply:
Yes, I think that makes sense as we will have this ad-hoc trait documented and well specified. Moreover, if `trait ToAdviceInputs: IntoIterator<Item = Felt>`, we can just `collect()` vectors as well

Reply:
Would we need to modify the return type / iterator to account for the fact we have three data sources - `advice_stack`, `advice_map`, and `merkle_store`?

Reply:
I'm not sure `IntoIterator` would be needed. One approach we could take is to have two traits: `AdviceInputs` and `ToAdviceInputs` - these traits could look as follows:
```Rust
pub trait AdviceInputs {
    fn push_onto_stack(&mut self, values: &[Felt]);
    fn insert_into_map(&mut self, key: Word, values: Vec<Felt>);
    fn add_merkle_path(&mut self, path: MerklePath);
}

pub trait ToAdviceInputs {
    fn to_advice_inputs<T: AdviceInputs>(&self, target: &mut T);
}
```

Then, for example, we could implement `AdviceInputs` on `MemAdviceProvider` and `ToAdviceInputs` on various structs.

One drawback of the above approach is that sometimes how we reduce an object to advice inputs may be contextual - i.e., in some cases the data could go into advice map but in other cases it could go onto advice stack. The above approach wouldn't be able to handle it. But I'm not sure if that's a big deal.

Reply:
Another approach could be to have a single trait with 3 methods:
```Rust
pub trait ToAdviceInputs {
    fn get_stack_inputs(&self) -> Vec<Felt>;
    fn get_map_inputs(&self) -> BTreeMap<[u8; 32], Vec<Felt>>>;
    fn get_merkle_paths(&self) -> Vec<MerklePath>;
}
```
But I'm not sure if it is better than the first approach.

Reply:
I think we should extract this into an issue. Conclude on a standard (I think @bobbinth proposal makes sense) and then sweep all repos to implement the trait where required. I'll go ahead and do that.